### ISSUE TITLE
single timestep prediction using LSTM

### ISSUE THREAD

**Issue by vendra** [keras-team/keras#8771]
I am not sure how the prediction works with LSTM.
I can correctly train a model using this code:

```
model.add(Masking(mask_value= -1.0, input_shape=(None, 5)))
model.add(LSTM(units=nr_units, return_sequences=True, activation='relu'))
model.add(Dropout(0.2))
model.add(TimeDistributed(Dense(6, activation='sigmoid')))
model.compile(loss="categorical_crossentropy", optimizer='adam', metrics=['accuracy'])
model.fit(x_train, y_train, epochs=nr_epochs, batch_size=nr_batch, validation_split=0.2)
```

My question is: can I predict the output of every timestep in a sequence by providing only one timestep as input? I can't do otherwise because only the first element of the sequence is known while the 2,..,n depends on output of the prediction.

I am making the prediction like this inside a loop:
`pred = model.predict(x_test)
`
Where x_test is a single 'frame' of a sequence. Does the model retains its internal state like this? Or do I need to provide the full input sequence?

Thanks

**Comment by mturnshek** [comment]
Can you provide a fuller explanation of what you're trying to accomplish? Are you trying to train in batch, and predict one frame at a time, such that the LSTM state is retained between predictions?

**Comment by vendra** [comment]
@mturnshek Yes. During training time I can provide the full input and output sequences. But at test time I can't provide a full input sequence, because the frame of the input depends on the previous output.

Like you said, I'd like to be able to predict each timesteps using separates invocations of predict() maintining the state between the predictions.

**Comment by David-Murray** [comment]
If I understand your input is for example, x1, x2, x3 with output y1 which during training you have all of the data however during prediction you have:
x1, x2, x3 = y1; First has all known variable
x2, x3, y1 = y2; Each subsequent prediction needs the previous prediction
x3, y1, y2 = y3;
Is that correct?

**Comment by vendra** [comment]
@David-Murray No, I have an encoded representation of the environment that depends on the previous output of the network.

I want to be able to feed to the network the first environment observation (which has 5 features). After this observation I need a prediction (6 features) that has some effects on the next environment observation. The problem is I can't provide the full input sequence a priori. I want to be able to call predict() at each timestep and compute the new input in between those calls. 

So now I can feed the new observation and get the new prediction and so on. 
![lstm-core](https://user-images.githubusercontent.com/1921364/33943682-0fddc582-e01a-11e7-876d-41af3bf6dd46.png)

**Comment by mturnshek** [comment]
To train in batch, but predict in sequence, you can make two models which are exactly the same, but have different ‘batch_input_shapes’.

After training the first model in batch, you transfer the weights to the second model to predict.

Make the training model with the first dimension in batch_input_shape as your training batch size. Then, make a new prediction model which is the same, except the first dimension in batch_input_shape is 1.

In both models, set stateful=True, and return_sequences=True.

You can then use model.reset_states to reset your recurrent states during the appropriate times during training or predicting.

Here’s some code that might help.
https://github.com/mturnshek/deep-learning/blob/master/realtime_rnn_predictions/batch_train_realtime_predict.py

**Comment by mpariente** [comment]
Hey @mturnshek,  the [example](https://github.com/mturnshek/deep-learning/blob/master/realtime_rnn_predictions/batch_train_realtime_predict.py) you linked to looks nice. 

I agree that to make real-time predictions while keeping the states, you need to set the prediction model as `stateful`, however, it doesn't seem needed in the training model. Maybe it was needed in your example but I doesn't appear to be necessary. Am I missing something or do you agree?

**Comment by mturnshek** [comment]
@mpariente My example is just one way, but I'm sure there are other ways of doing it. Here are my thoughts on why it's that way in the example I provided.

Keep in mind the input shape for a GRU is `(batch_size, timesteps, input_dim)`. So the shape of `x` and `y` in the example are `(600900, 1, 59)`. So 600900 data points, 1 timestep, and 59 possible characters. Since in my example there is only one timestep each data point, setting `stateful=False` on the training model would make it so that each timestep would be treated as an individual case and so it would learn to predict the most likely character following another with no recurrence.

By setting `stateful=True` and resetting the states every epoch, it trains from the beginning to end of the book, then resets for the next epoch. Notice `shuffle=False` on the data so it retains order.

**Comment by lmxhappy** [comment]
Supposed that past 7 days' history data influences the current data. A model is trained and its timestep = 3. When predicting if one timestep of x1, x2, x3   is passed to the model, the result is not satisfying.
 @mturnshek hi, do you mean that the state that is produced by one predict call is used by the next call?  If yes, for the targeted prediction, more sequences and more time steps are passed to the model and get more predictions although one the last prediction is targeted.
